internal/cortex/frontend/transport/handler.go

```diff
@@ -122,14 +165,88 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	r.Body = http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize)
 	r.Body = io.NopCloser(io.TeeReader(r.Body, &buf))
 
+	// Check if caching is enabled
+	if f.lru != nil {
+		// Store query expression
+		queryExpressionNormalized = f.regex.ReplaceAllString(r.URL.Query().Get("query"), " ")
+
+		// Time range length for queries, if either of "start" or "end" are not present, return 0
+		getQueryRangeSeconds := func() int {
+			start, err := strconv.Atoi(r.URL.Query().Get("start"))
+			if err != nil {
+				return 0
+			}
+			end, err := strconv.Atoi(r.URL.Query().Get("end"))
+			if err != nil {
+				return 0
+			}
+			return end - start
+		}
+
+		// Store query time range length
+		queryExpressionRangeLength = getQueryRangeSeconds()
+
+		// Check if query in cache and whether value exceeds time range length
+		if value, ok := f.lru.Get(queryExpressionNormalized); ok && value.(int) >= queryExpressionRangeLength {
+			w.WriteHeader(http.StatusForbidden)
+			level.Warn(util_log.WithContext(r.Context(), f.log)).Log("msg", "FOUND QUERY IN CACHE: CAUSED ERROR: ", "query expression", queryExpressionNormalized)
+			f.cachedHits.Inc()
+			return
+		}
+	}
+
 	startTime := time.Now()
 	resp, err := f.roundTripper.RoundTrip(r)
 	queryResponseTime := time.Since(startTime)
 
 	if err != nil {
 		writeError(w, err)
+		queryString = f.parseRequestQueryString(r, buf)
+
+		// Check if caching is enabled
+		if f.lru != nil {
+			// Extracting error code
+			codeExtract := f.errorExtract.FindStringSubmatch(err.Error())
+
+			// Checking if error code extracted successfully
+			if codeExtract != nil && len(codeExtract) >= 2 {
+
+				// Converting error code to int
+				errCode, strConvError := strconv.Atoi(codeExtract[1])
+
+				// Checking if error code extracted properly
+				if strConvError != nil {
+					level.Error(util_log.WithContext(r.Context(), f.log)).Log(
+						"msg", "String to int conversion error", "response ", resp,
+					)
+				}
+
+				// If error should be cached, store it in cache
+				if isCacheableError(errCode) {
+					//checks if queryExpression is already in cache, and updates time range length value if it is shorter
+					if contains, _ := f.lru.ContainsOrAdd(queryExpressionNormalized, queryExpressionRangeLength); contains {
+						if oldValue, ok := f.lru.Get(queryExpressionNormalized); ok {
+							queryExpressionRangeLength = min(queryExpressionRangeLength, oldValue.(int))
+						}
+						f.lru.Add(queryExpressionNormalized, queryExpressionRangeLength)
+					}
+
+					level.Info(util_log.WithContext(r.Context(), f.log)).Log(
+						"msg", "Cached query due to cacheable error code", "response ", resp,
```

Comment:
these logs level should be Debug, make sure logs are not too verbose to hide actual important ones

Reply:
changed log level and made logs more concise